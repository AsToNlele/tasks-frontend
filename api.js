import axios from 'axios';
import {
  TASKS_API_ROOT,
  AVAILABLE_TASKS_ROOT,
  EXECUTED_TASKS_ROOT,
  SYSTEMS_ROOT,
} from './src/constants';

const getTasks = async (path) => {
  let response;

  const request = await axios
    .get(TASKS_API_ROOT.concat(path))
    .catch(function (error) {
      return error;
    });

  if (request.status === 200) {
    response = request.data;
  } else {
    response = request;
  }

  return response;
};

export const fetchAvailableTasks = () => {
  return getTasks(AVAILABLE_TASKS_ROOT);
};

export const fetchAvailableTask = (slug) => {
  return getTasks(AVAILABLE_TASKS_ROOT.concat(`/${slug}`));
};

export const fetchExecutedTasks = () => {
  return getTasks(EXECUTED_TASKS_ROOT);
};

export const fetchExecutedTask = (path) => {
  return getTasks(EXECUTED_TASKS_ROOT.concat(path));
};

export const fetchExecutedTaskJobs = (id) => {
  let idPath = `/${id}`;
  return fetchExecutedTask(idPath.concat('/jobs'));
};

export const fetchSystems = (path) => {
  return getTasks(SYSTEMS_ROOT.concat(path));
};
