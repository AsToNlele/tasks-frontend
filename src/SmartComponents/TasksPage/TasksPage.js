import React, { Suspense, lazy, useState } from 'react';

import { Spinner, StackItem, Stack } from '@patternfly/react-core';
import { Main } from '@redhat-cloud-services/frontend-components/Main';
import {
  PageHeader,
  PageHeaderTitle,
} from '@redhat-cloud-services/frontend-components/PageHeader';
import { TasksTabs } from '../../PresentationalComponents';
import { TASKS_PAGE_TABS } from '../../constants';

const AvailableTasks = lazy(() => import('../AvailableTasks/AvailableTasks'));
const CompletedTasks = lazy(() => import('../CompletedTasks/CompletedTasks'));

import './tasks-page.scss';

const TasksPage = () => {
  const [tabIndex, setTab] = useState(0);

  const updateTab = (event, index) => {
    setTab(index);
  };

  return (
    <React.Fragment>
      <PageHeader>
        <PageHeaderTitle title="Tasks" />
      </PageHeader>
      <TasksTabs
        className="tabs-background"
        tabIndex={tabIndex}
        updateTab={updateTab}
        tabsList={TASKS_PAGE_TABS}
      />
      <Main>
        <Stack hasGutter>
          <StackItem>
            <Suspense fallback={<Spinner />}>
              {tabIndex === 0 ? <AvailableTasks /> : <CompletedTasks />}
            </Suspense>
          </StackItem>
        </Stack>
      </Main>
    </React.Fragment>
  );
};

export default TasksPage;
