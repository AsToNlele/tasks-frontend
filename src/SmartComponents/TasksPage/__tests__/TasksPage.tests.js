import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import TasksPage from '../TasksPage';
import { fetchExecutedTasks } from '../../../../api';
import { fetchAvailableTasks } from '../../../../api';

jest.mock('../../../../api');

describe('TasksPage', () => {
  let props;
  let mockStore = configureStore();

  it('should render correctly', async () => {
    fetchAvailableTasks.mockImplementation(async () => {
      return { data: [] };
    });
    const store = mockStore(props);
    const { asFragment } = render(
      <MemoryRouter keyLength={0}>
        <Provider store={store}>
          <TasksPage tab={0} />
        </Provider>
      </MemoryRouter>
    );

    await waitFor(() => expect(asFragment()).toMatchSnapshot());
    await waitFor(() =>
      expect(screen.getByLabelText('available-tasks')).toBeInTheDocument()
    );
  });

  it('should update tab index', async () => {
    fetchExecutedTasks.mockImplementation(async () => {
      return { meta: { count: 0 }, data: [] };
    });
    const store = mockStore(props);

    render(
      <MemoryRouter keyLength={0}>
        <Provider store={store}>
          <TasksPage {...props} />
        </Provider>
      </MemoryRouter>
    );

    await userEvent.click(screen.getByText('Completed tasks'));
    await waitFor(() =>
      expect(screen.getByLabelText('completed-tasks')).toBeInTheDocument()
    );
    await userEvent.click(screen.getByText('Available tasks'));
    await waitFor(() =>
      expect(screen.getByLabelText('available-tasks')).toBeInTheDocument()
    );
  });
});
