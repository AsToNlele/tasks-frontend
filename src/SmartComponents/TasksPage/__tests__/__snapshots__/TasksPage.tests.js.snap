// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksPage should render correctly 1`] = `
<DocumentFragment>
  <section
    class="pf-l-page-header pf-c-page-header pf-l-page__main-section pf-c-page__main-section pf-m-light"
    widget-type="InsightsPageHeader"
  >
    <h1
      class="pf-c-title pf-m-2xl"
      data-ouia-component-id="OUIA-Generated-Title-1"
      data-ouia-component-type="PF4/Title"
      data-ouia-safe="true"
      widget-type="InsightsPageHeaderTitle"
    >
      Tasks
    </h1>
  </section>
  <div
    class="pf-c-tabs tabs-background"
    data-ouia-component-id="OUIA-Generated-Tabs-1"
    data-ouia-component-type="PF4/Tabs"
    data-ouia-safe="true"
  >
    <button
      aria-hidden="true"
      aria-label="Scroll left"
      class="pf-c-tabs__scroll-button"
      disabled=""
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 256 512"
        width="1em"
      >
        <path
          d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
        />
      </svg>
    </button>
    <ul
      class="pf-c-tabs__list"
    >
      <li
        class="pf-c-tabs__item pf-m-current"
      >
        <button
          aria-controls="pf-tab-section-0-tabs-page-0"
          class="pf-c-tabs__link"
          data-ouia-component-type="PF4/TabButton"
          data-ouia-safe="true"
          id="pf-tab-0-tabs-page-0"
        >
          <span
            class="pf-c-tabs__item-text"
          >
            Available tasks
          </span>
        </button>
      </li>
      <li
        class="pf-c-tabs__item"
      >
        <button
          aria-controls="pf-tab-section-1-tabs-page-1"
          class="pf-c-tabs__link"
          data-ouia-component-type="PF4/TabButton"
          data-ouia-safe="true"
          id="pf-tab-1-tabs-page-1"
        >
          <span
            class="pf-c-tabs__item-text"
          >
            Completed tasks
          </span>
        </button>
      </li>
    </ul>
    <button
      aria-hidden="true"
      aria-label="Scroll right"
      class="pf-c-tabs__scroll-button"
      disabled=""
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 256 512"
        width="1em"
      >
        <path
          d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
        />
      </svg>
    </button>
  </div>
  <section
    class="pf-l-page__main-section pf-c-page__main-section"
    page-type=""
  >
    <div
      class="pf-l-stack pf-m-gutter"
    >
      <div
        class="pf-l-stack__item"
      >
        <span
          aria-label="Contents"
          aria-valuetext="Loading..."
          class="pf-c-spinner pf-m-xl"
          role="progressbar"
        >
          <span
            class="pf-c-spinner__clipper"
          />
          <span
            class="pf-c-spinner__lead-ball"
          />
          <span
            class="pf-c-spinner__tail-ball"
          />
        </span>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
